File: dart/glue/lib/src/lib/bool/fallback.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

final Ir fallback = IrSpecial(fallbackImpl);

Eval<Ir> fallbackImpl(List<Ir> args) {
  return switch (args) {
    [final a, final b] => eval(a).bind((a_) {
      return isExist(a_) ? Eval.pure(a_) : eval(b);
    }),
    _ => throwError(wrongNumberOfArguments()),
  };
}

File: haskell/glue/src/Glue/Lib/Bool/Fallback.hs
module Glue.Lib.Bool.Fallback where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception (wrongNumberOfArguments)
import Glue.IR (IR (..), exists)

fallback :: IR Eval
fallback = Special fallbackImpl

fallbackImpl :: [IR Eval] -> Eval (IR Eval)
fallbackImpl [a, b] = do
    a' <- eval a
    if exists a'
        then pure a'
        else eval b
fallbackImpl _ = throwError wrongNumberOfArguments
